File: csrc/random.c
// Copyright (c) 2026 渟雲. All rights reserved.
#include "random.h"
#include <intrin.h>

static ULONGLONG s_state[4];

static __forceinline ULONGLONG Rotl64(ULONGLONG x, int k) {
  return (x << k) | (x >> (64 - k));
}

static __forceinline ULONGLONG SplitMix64(ULONGLONG* x) {
  ULONGLONG z = (*x += 0x9E3779B97F4A7C15ULL);
  z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9ULL;
  z = (z ^ (z >> 27)) * 0x94D049BB133111EBULL;
  return z ^ (z >> 31);
}

static __forceinline ULONGLONG Xoshiro256StarStar(void) {
  ULONGLONG* s = s_state;
  ULONGLONG result = Rotl64(s[1] * 5, 7) * 9;
  ULONGLONG t = s[1] << 17;

  s[2] ^= s[0];
  s[3] ^= s[1];
  s[1] ^= s[2];
  s[0] ^= s[3];
  s[2] ^= t;
  s[3] = Rotl64(s[3], 45);

  return result;
}

static __forceinline ULONGLONG ReadInterruptTimeRaw(void) {
  volatile KSYSTEM_TIME* t = &SharedUserData->InterruptTime;
  ULONG lo = t->LowPart;
  LONG hi1 = t->High1Time;
  LONG hi2 = t->High2Time;
  if (hi1 != hi2) {
    hi1 = t->High1Time;
  }
  return (((ULONGLONG)(ULONG)hi1) << 32) | lo;
}

// Entropy from：
// __rdtsc()
// SharedUserData
// STACK
VOID RandomEngineInit(VOID) {
  ULONGLONG entropy[4];

  entropy[0] = __rdtsc();
  entropy[1] = ReadInterruptTimeRaw();
  entropy[2] = (ULONGLONG)(ULONG_PTR)&entropy;
  entropy[3] = __rdtsc();
  ULONGLONG mix = 0x243F6A8885A308D3ULL;
  for (int i = 0; i < 4; i++) {
    mix ^= entropy[i];
    s_state[i] = SplitMix64(&mix);
  }

  if ((s_state[0] | s_state[1] | s_state[2] | s_state[3]) == 0) {
    s_state[0] = 0x9E3779B97F4A7C15ULL;
    s_state[1] = 0xD1B54A32D192ED03ULL;
    s_state[2] = 0x85E1C3D753D46D27ULL;
    s_state[3] = 0x94D049BB133111EBULL;
  }
}

ULONGLONG RandomEngineNext(VOID) { return Xoshiro256StarStar(); }

File: src/random.rs
// Copyright (c) 2026 渟雲. All rights reserved.

const KUSER_SHARED_INTERRUPT_TIME: u64 = 0xFFFF_F780_0000_0008;

static mut S_STATE: [u64; 4] = [0; 4];

#[inline]
fn rotl64(x: u64, k: u32) -> u64 {
    (x << k) | (x >> (64 - k))
}

#[inline]
fn splitmix64(x: &mut u64) -> u64 {
    *x = x.wrapping_add(0x9E37_79B9_7F4A_7C15);
    let mut z = *x;
    z = (z ^ (z >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
    z = (z ^ (z >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
    z ^ (z >> 31)
}

#[inline]
fn xoshiro256_starstar() -> u64 {
    unsafe {
        let s = &mut *core::ptr::addr_of_mut!(S_STATE);
        let result = rotl64(s[1].wrapping_mul(5), 7).wrapping_mul(9);
        let t = s[1] << 17;

        s[2] ^= s[0];
        s[3] ^= s[1];
        s[1] ^= s[2];
        s[0] ^= s[3];
        s[2] ^= t;
        s[3] = rotl64(s[3], 45);

        result
    }
}

pub fn random_engine_next() -> u64 {
    xoshiro256_starstar()
}

#[inline]
fn read_interrupt_time_raw() -> u64 {
    let ptr = KUSER_SHARED_INTERRUPT_TIME as *const u32;
    let lo = unsafe { *ptr };
    let mut hi1 = unsafe { *ptr.add(1) } as i32;
    let hi2 = unsafe { *ptr.add(2) } as i32;
    if hi1 != hi2 {
        hi1 = unsafe { *ptr.add(1) } as i32;
    }
    ((hi1 as u32 as u64) << 32) | lo as u64
}

fn rdtsc() -> u64 {
    let mut lo: u32;
    let mut hi: u32;
    unsafe {
        core::arch::asm!("rdtsc", out("eax") lo, out("edx") hi, options(nostack));
    }
    ((hi as u64) << 32) | lo as u64
}

pub fn random_engine_init() {
    unsafe {
        let mut entropy = [0u64; 4];
        entropy[0] = rdtsc();
        entropy[1] = read_interrupt_time_raw();
        entropy[2] = core::ptr::addr_of!(entropy) as u64;
        entropy[3] = rdtsc();

        let mut mix = 0x243F_6A88_85A3_08D3u64;
        for i in 0..4 {
            mix ^= entropy[i];
            S_STATE[i] = splitmix64(&mut mix);
        }

        if (S_STATE[0] | S_STATE[1] | S_STATE[2] | S_STATE[3]) == 0 {
            S_STATE[0] = 0x9E37_79B9_7F4A_7C15;
            S_STATE[1] = 0xD1B5_4A32_D192_ED03;
            S_STATE[2] = 0x85E1_C3D7_53D4_6D27;
            S_STATE[3] = 0x94D0_49BB_1331_11EB;
        }
    }
}
